### PR TITLE
Default opacity to 1 when fill color is set

### PR DESCRIPTION
**Pull request by ryanmccauley211**
## Changelog / Overview

Enhancement for #1682

Set fill opacity to 1 by default when fill color is set

## Motivation

When fill color is set it doesn't make sense to default opacity to 0, as the user would expect to see the fill color

## Explanation for Changes

Makes it more intuitive setting colors

## Testing Status

I needed to change the default opacity to None to handle 0 opacity from opacity that is not set. This is because svgs have their own defaults for opacity and fill color and this change would change that behaviour otherwise

## Checklist
- [X] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [X] I have written a descriptive PR title (see top of PR template for examples)
- [X] I have written a changelog entry for the PR or deem it unnecessary
- [X] My new functions/classes either have a docstring or are private
- [X] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [X] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

File: tests/test_vectorized_mobject.py [hunk 1]
```diff
@@ -204,3 +204,24 @@ def test_vgroup_item_assignment_only_allows_vmobjects():
     vgroup = VGroup(VMobject())
     with pytest.raises(TypeError, match="All submobjects must be of type VMobject"):
         vgroup[0] = "invalid object"
+
+
+def test_correct_fill_opacity_default_when_fill_set():
+    """Test opacity defaults to 1 when fill_color is set and opacity is not set"""
```
Comment (hydrobeam) [inline tests/test_vectorized_mobject.py:210]:
```suggestion
    """Test that opacity defaults to 1 when ``fill_color`` is set and opacity is not set."""
```

File: tests/test_vectorized_mobject.py [hunk 2]
```diff
@@ -204,3 +204,24 @@ def test_vgroup_item_assignment_only_allows_vmobjects():
...
+    vmobject = VMobject(fill_color="#FF0000")
+    expected_opacity = 1.0
+    assert vmobject.fill_opacity == expected_opacity
+
+
+def test_correct_fill_opacity_default_when_fill_not_set():
+    """Test opacity defaults to 0 when fill_color is not set and opacity is not set"""
```
Comment (hydrobeam) [inline tests/test_vectorized_mobject.py:217]:
```suggestion
    """Test that opacity defaults to 0 when ``fill_color`` is not set and opacity is not set."""
```

File: tests/test_vectorized_mobject.py [hunk 3]
```diff
@@ -204,3 +204,24 @@ def test_vgroup_item_assignment_only_allows_vmobjects():
...
+    vmobject = VMobject()
+    expected_opacity = 0.0
+    assert vmobject.fill_opacity == expected_opacity
+
+
+def test_fill_opacity_does_not_default_when_set():
+    """Test opacity defaults to 1 when fill_color is set and opacity is not set"""
```
Comment (hydrobeam) [inline tests/test_vectorized_mobject.py:224]:
```suggestion
    """Test that opacity is correctly processed when ``fill_color`` and ``fill_opacity`` are set."""
```

**Review by hydrobeam — APPROVED** [review]
Left some adjustments to the wording on the test documentation, but LGTM

**Comment by behackl** [comment]
There was a lengthy discussion about this (or a very much related issue, didn't compare details) in #956, where it was decided not to implement this change.

Personally I feel that a more intelligently behaving fill_opacity makes sense, but I don't like that the behavior of `set_color` (which calls `set_stroke` and `set_fill`) is also changed with this PR.

**Comment by ryanmccauley211** [comment]
> There was a lengthy discussion about this (or a very much related issue, didn't compare details) in #956, where it was decided not to implement this change.
> 
> Personally I feel that a more intelligently behaving fill_opacity makes sense, but I don't like that the behavior of `set_color` (which calls `set_stroke` and `set_fill`) is also changed with this PR.

Ah I didn't realise there was a discussion about this before. So the scenario where it would affect set_color is when a user initially sets the fill, it then flips from defaulting to 0 to 1 and then it will remain at 1 when calling set_color? That behaviour seems okay to me if the user has already intended to set a fill, but I read through the other discussion and I guess there is a differing opinions on it. I guess my PR is also incomplete then because I would need to do the same for set_fill for it to be consistent

My personal experience was similar to the @sparshg who originally created the other PR. When I started using manim I was confused as to why set_fill didn't seem to be doing anything

**Comment by behackl** [comment]
I've now also read your code (which I should have done before, apologies): your approach and what was proposed in #956 are certainly different. I was worried that this PR changes the behavior of `set_color` such that whenever the fill opacity was not explicitly set to 0, `set_color` would also set the fill opacity to 1 -- such that `Square().set_color(RED)` would have produced a square with red fill.

I think that the behavior you propose here is more user friendly: after all, the case that someone *explicitly* specifies a `fill_color` upon mobject creation but simultaneously wants a fill opacity of 0 is rather strange. It is natural to expect that when a fill color is set, the mobject will actually be colored that way -- and on top of that, the old behavior can still be restored by setting `fill_opacity=0`.

Assuming this PR doesn't introduce problems with SVG parsing (where, if necessary the old behavior could be kept by setting `fill_opacity=0` in `SVGMobject.__init__`), I think this is a good idea.

**Comment by sparshg** [comment]
@behackl, my PR didn't change the behavior of `set_color()`, It has exactly same behavior as now, nothing changed, just added optional opacity parameter if you want to adjust both together.

The only things my PR did is
`set_fill/set_stroke` "If the opacity parameter is not passed, AND the fill/stroke is transparent, make the fill/stroke opaque"

which still makes absolute sense to me, why would you even fill color in transparent object without passing opacity.
This PR also does the same thing, but only during object initialization.
I still don't know why my PR was closed (except that it became a bit messy)

**Comment by ryanmccauley211** [comment]
> Assuming this PR doesn't introduce problems with SVG parsing (where, if necessary the old behavior could be kept by setting `fill_opacity=0` in `SVGMobject.__init__`), I think this is a good idea.

I couldn't see any issues in svgs, they have the opacity defaulting to 1 and fill for paths at least seem to default to black. I did some testing with svgs but am mostly relying on the existing tests. Do you think having this same behaviour in set_fill is needed?

**Review by huguesdevimeux — APPROVED** [review]

**Comment by huguesdevimeux** [comment]
I think that's fine, and I actually don't see how it could produce any issue since when `fill_opacity` is `None`, [manim interprets it  anyway as a 0.](https://github.com/ManimCommunity/manim/blob/723e645abce9c86cdf40c98959157e6ff93896a0/manim/mobject/types/vectorized_mobject.py#L164-L165)

**Comment by ryanmccauley211** [comment]
> I think that's fine, and I actually don't see how it could produce any issue since when `fill_opacity` is `None`, [manim interprets it anyway as a 0.](https://github.com/ManimCommunity/manim/blob/723e645abce9c86cdf40c98959157e6ff93896a0/manim/mobject/types/vectorized_mobject.py#L164-L165)

Ok cool, and do you think this behaviour should also be extended to set_fill()? This PR currently only does it on initialisation

**Comment by huguesdevimeux** [comment]
> > I think that's fine, and I actually don't see how it could produce any issue since when `fill_opacity` is `None`, [manim interprets it anyway as a 0.](https://github.com/ManimCommunity/manim/blob/723e645abce9c86cdf40c98959157e6ff93896a0/manim/mobject/types/vectorized_mobject.py#L164-L165)
> 
> Ok cool, and do you think this behaviour should also be extended to set_fill()? This PR currently only does it on initialisation

I haven't thought of that, but sure it should be. The best way would be to create a property `fill_color` and to ensure that opacity is not `None` when the color is changed.

**Comment by huguesdevimeux** [comment]
Maybe for another PR ?

**Comment by hydrobeam** [comment]
So is this PR good to merge? Seems like it's stalled a bit...

**Comment by Darylgolden** [comment]
> So is this PR good to merge? Seems like it's stalled a bit...

Isn't there some consensus not to do it?

**Comment by ryanmccauley211** [comment]
> > So is this PR good to merge? Seems like it's stalled a bit...
> 
> Isn't there some consensus not to do it?

Don’t think there’s a consensus either way tbh

**Comment by ryanmccauley211** [comment]
> So is this PR good to merge? Seems like it's stalled a bit...

I’m still in favour of taking this PR a little further to add this behaviour to the setters as well as the constructor, maybe we should open a discussion on discord to come to an agreement

**Comment by behackl** [comment]
Yes, I think we should have a (GitHub?) discussion about this, and it would be good to hear the opinion of both the core devs and the community.

Here is an attempt at boiling this situation down to two questions we might be able to have a meaningful discussion over:

- From a UI/UX point of view, does it make sense that setting some property also changes the values of other mobject properties? Could there be cases where this behavior is not desireable? And if so, would, say, an INFO log entry be sufficient to make a user aware of this?
- Is the situation for fill colors / opacity *special* enough to justify breaking the paradigm that `set_property`-methods should **only** interact with the property they are setting?

**Comment by behackl** [comment]
I'll close this PR. As mentioned in my previous comment, we should have a discussion about this, and if we decide to go through with this, change it consistently.

**Pull request closed without merging**